Sources/KSCrashRecordingCore/KSDate.c

```diff
@@ -25,32 +25,48 @@
 #include "KSDate.h"
 
 #include <stdio.h>
+#include <string.h>
 #include <sys/time.h>
 #include <time.h>
 
-void ksdate_utcStringFromTimestamp(time_t timestamp, char *buffer21Chars)
+void ksdate_utcStringFromTimestamp(time_t timestamp, char *buffer, size_t bufferSize)
 {
+    memset(buffer, 0, bufferSize);
+
+    // Cannot be in the future.
+    if ((uint64_t)timestamp > ksdate_seconds()) {
+        return;
+    }
+
     struct tm result = { 0 };
-    gmtime_r(&timestamp, &result);
-    snprintf(buffer21Chars, 21, "%04d-%02d-%02dT%02d:%02d:%02dZ", result.tm_year + 1900, result.tm_mon + 1,
+    if (!gmtime_r(&timestamp, &result)) {
+        return;
+    }
+    snprintf(buffer, bufferSize, "%04d-%02d-%02dT%02d:%02d:%02dZ", result.tm_year + 1900, result.tm_mon + 1,
              result.tm_mday, result.tm_hour, result.tm_min, result.tm_sec);
 }
 
-void ksdate_utcStringFromMicroseconds(int64_t microseconds, char *buffer28Chars)
+void ksdate_utcStringFromMicroseconds(int64_t microseconds, char *buffer, size_t bufferSize)
 {
+    memset(buffer, 0, bufferSize);
+
+    // Cannot be in the future.
+    if ((uint64_t)microseconds >= ksdate_microseconds()) {
+        return;
+    }
+
     struct tm result = { 0 };
     time_t curtime = (time_t)(microseconds / 1000000);
     long micros = (long)(microseconds % 1000000);
 
-    gmtime_r(&curtime, &result);
-    snprintf(buffer28Chars, 28, "%04d-%02d-%02dT%02d:%02d:%02d.%06ldZ", result.tm_year + 1900, result.tm_mon + 1,
+    if (!gmtime_r(&curtime, &result)) {
+        return;
+    }
+
+    snprintf(buffer, bufferSize, "%04d-%02d-%02dT%02d:%02d:%02d.%06ldZ", result.tm_year + 1900, result.tm_mon + 1,
              result.tm_mday, result.tm_hour, result.tm_min, result.tm_sec, micros);
 }
 
-uint64_t ksdate_microseconds(void)
-{
-    struct timeval tp;
-    gettimeofday(&tp, NULL);
```

Comment:
I'd prefer if we kept using `gettimeofday` since it's portable. I haven't abandoned my hopes of porting this to Android ;-)

Reply:
From what I gather, it’s not safe to use in a signal handler.

Reply:
`gettimeofday` only locks if the timezone field is set (https://github.com/apple-oss-distributions/xnu/blob/main/bsd/kern/kern_time.c#L150); otherwise it only does a syscall (https://github.com/apple-oss-distributions/xnu/blob/main/bsd/kern/kern_time.c#L129) and copies out the data.

Also, the tz argument is no longer valid, and must be set to NULL now (https://man.openbsd.org/gettimeofday)